### PR TITLE
Handle table cell backgrounds during display list generation for <table>

### LINKED ISSUE #19788: <tr> backgrounds aren't drawn correctly
Other browsers seem to render `tr { background-color: black }` the same as  `tr > th, tr > td { background-color: black }`.  Servo does not. (Not sure where that's spec'ed, or even if it is)

Testcase: http://kwan.perix.co.uk/mozilla/tableRowBackgroundTest.html
(Correct rendering emulated on hover using cell backgrounds)

Servo rendering:
screenshot from 2018-01-16 22-50-32

Firefox rendering:
screenshot from 2018-01-16 22-51-31

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #19788

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20034)

### CODE REVIEW

**Comment by Manishearth** [comment]
Okay, I've gotten something working. Tables with styled columns/colgroups/rowgroups/rows/tables, that use colspan work.

There are a couple of missing pieces:
 
 - I don't handle currentColor correctly
 - I don't support rowspan (yet)
 - I no longer call build_display_list in TableRowFlow -- this may cause borders to be skipped? I'm not sure about this
 - Backgrounds are positioned relative to the cell, not the region being styled


For the last one, it's a bit tricky. I can handle it for row/rowgroup backgrounds by using the stackingcontext-relative position of the element for drawing the background, and then clipping it with the regular element's position.

But for columns/colgroups, we don't store the positions anywhere. We kinda calculate these during border-collapse layout; so it might be worth adding a layout pass to compute these and store them on ColGroupFlow.

Thoughts?

**Comment by Manishearth** [comment]
Oh, also, a lot of this is made complicated by the fact that build_display_list needs a mutable reference, but looking through the code that may not actually be necessary.

**Comment by mbrubeck** [comment]
> But for columns/colgroups, we don't store the positions anywhere. We kinda calculate these during border-collapse layout; so it might be worth adding a layout pass to compute these and store them on ColGroupFlow.

Yes, seems like this will be necessary (though this can be a follow-up if you want).

> Oh, also, a lot of this is made complicated by the fact that build_display_list needs a mutable reference, but looking through the code that may not actually be necessary.

Yeah... many of the individual builder methods only take`&self`.  Some like `build_fragment_type_specific_display_items` take `&mut self` but don't actually need it. It should definitely be possible to do most of this without `&mut` access.

Does `Fragment::build_display_list` only use `mut` to unset `self.restyle_damage`?  If so, maybe we could separate the rest of its body into an `&self` helper function.

> I no longer call build_display_list in TableRowFlow -- this may cause borders to be skipped?

Yes, this will need to be addressed. I haven't looked at your code in detail yet, but it seems there's not a clear place to fit this in, since I think row borders should be drawn after the backgrounds but before the cell contents?  Or does this even matter – is it possible for the borders and backgrounds to overlap?  If not, then we could draw the row borders in separate loop iterating over the rows...

**Comment by Manishearth** [comment]
> Does Fragment::build_display_list only use mut to unset self.restyle_damage?

Yes.

I can handle this by explicitly doing the damage (and nothing else) in TableCellFlow's build_display_list

**Comment by Manishearth** [comment]
> Or does this even matter – is it possible for the borders and backgrounds to overlap?

I think not -- _however_, it is possible for the border and content to overlap since content can spill out. Hmm.

**Comment by Manishearth** [comment]
Made stuff immutable which lets me handle currentcolor easily

**Comment by Manishearth** [comment]
@bors-servo try

let's see what we get

**Comment by Manishearth** [comment]
Rowspan "done".

Given that we don't actually render rowspans where we should, it's a bit lackluster. But if we did, they would have working backgrounds.

**Comment by Manishearth** [comment]
Turns out borders are not skipped; table rows don't draw collapsed borders anyway, they delegate to cells!

I'm pushing bg positioning to a different bug.

There's one missing bit -- restyle damage doesn't get correctly applied here. I guess I should propagate it up in layout/incremental.rs?

**Comment by Manishearth** [comment]
So we fail css/cssom-view/elementFromPoint-dynamic-anon-box.html . I thought this was due to restyle damage, but I tried fixing that in incremental.rs and this still occurs.

File: components/layout/table_row.rs [hunk 1]
```diff
@@ -467,17 +467,8 @@ impl Flow for TableRowFlow {
...
-                                             .style
-                                             .get_inheritedtable()
-                                             .border_collapse {
-            BorderCollapse::Separate => BorderPaintingMode::Separate,
-            BorderCollapse::Collapse => BorderPaintingMode::Hidden,
-        };
-
-        self.block_flow.build_display_list_for_block(state, border_painting_mode);
+    fn build_display_list(&mut self, _: &mut DisplayListBuildState) { 
+        // handled in TableCellStyleInfo::build_display_list
```
Comment (mbrubeck) [inline components/layout/table_row.rs]: We'll still need to remove REPAINT damage here, or somewhere.

File: components/layout/table_cell.rs [hunk 2]
```diff
@@ -248,21 +247,16 @@ impl Flow for TableCellFlow {
...
-                                             .border_collapse {
-            BorderCollapse::Separate => BorderPaintingMode::Separate,
-            BorderCollapse::Collapse => BorderPaintingMode::Collapse(&self.collapsed_borders),
-        };
-
-        self.block_flow.build_display_list_for_block(state, border_painting_mode)
+        if self.visible {
+            // we skip setting the damage in TableCellStyleInfo::build_display_list()
+            // because we only have immutable access
+            self.block_flow.fragment.restyle_damage.remove(ServoRestyleDamage::REPAINT);
```
Comment (mbrubeck) [inline components/layout/table_cell.rs]: This should probably be unconditional.  (I think this was a bug in the old code.)

File: components/layout/table.rs [hunk 3]
```diff
@@ -867,3 +966,232 @@ impl<'a> Iterator for TableRowIterator<'a> {
...
+            None
+        }
+    }
+}
+
+impl<'table> TableCellStyleInfo<'table> {
+    fn build_display_list(&self, mut state: &mut DisplayListBuildState) {
+        if !self.cell.visible {
+            return
+        }
```
Comment (mbrubeck) [inline components/layout/table.rs]: I believe for empty cells (`cell.visible == false`), we should still draw the column and row backgrounds, just not the cell background and contents, per the spec: https://drafts.csswg.org/css2/tables.html#table-layers

File: components/layout/table.rs [hunk 4]
```diff
@@ -867,3 +966,233 @@ impl<'a> Iterator for TableRowIterator<'a> {
...
+    }
+}
+
+impl<'table> TableCellStyleInfo<'table> {
+    fn build_display_list(&self, mut state: &mut DisplayListBuildState) {
+        if !self.cell.visible || self.cell.block_flow.fragment.style()
+                                     .get_inheritedbox().visibility != Visibility::Visible {
+            return
+        }
```
Comment (mbrubeck) [inline components/layout/table.rs]: I believe for empty cells (cell.visible == false), we should still draw the column and row backgrounds, just not the cell background and contents, per the spec: https://drafts.csswg.org/css2/tables.html#table-layers

Comment (Manishearth) [reply components/layout/table.rs]: I tested this in Firefox, it doesn't work for columns at least. The cell disappears entirely.

Comment (Manishearth) [reply components/layout/table.rs]: Or rows. It only works for the surrounding table, and we still handle that

**Comment by mbrubeck** [comment]
@bors-servo r+

Thanks especially for factoring these traversals out into iterators. ❤️

**Comment by Manishearth** [comment]
@bors-servo r-

I"m still fixing tidy :)

**Comment by Manishearth** [comment]
@bors-servo retry

out of space

**Comment by Manishearth** [comment]
@bors-servo r=mbrubeck

oh, forgot to re-r

**Comment by Manishearth** [comment]
@bors-servo r=mbrubeck

display-012.xht passes on linux but fails on OSX, in Servo CI, Servo locally (macbook vs ubuntu), and Firefox locally, but for some reason Firefox doesn't have a metadata entry for it.

It's basically getting an extra pixel-height of background on the top and bottom (might be worth marking it as fuzzy), seems to be a font height issue?

**Comment by Manishearth** [comment]
@bors-servo retry

 - infra

**Comment by CYBAI** [comment]
@bors-servo retry

```
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/css/css-color/t421-rgb-values-meaning-b.xht", "line": 49133, "action": "test_result", "expected": "PASS"}
```

**Comment by Manishearth** [comment]
That test has a table with lots of per-cell backgrounds, perhaps this PR slows down tables because we build too many display lists? Trying an optimization to fix this.

**Comment by ferjm** [comment]
@bors-servo retry

```
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/WebCryptoAPI/idlharness.https.worker.html", "line": 147751, "action": "test_result", "expected": "OK"}
```

**Comment by CYBAI** [comment]
Error in [linux-rel-nogate](http://build.servo.org/builders/linux-rel-nogate/builds/1690/steps/shell__5/logs/stdio)

```
error[E0425]: cannot find value `handle` in this scope
    --> components/script/dom/xmlhttprequest.rs:1068:47
     |
1068 |             self.global().unschedule_callback(handle);
     |                                               ^^^^^^ not found in this scope

error: aborting due to previous error

error: Could not compile `script`.
```

**Comment by Manishearth** [comment]
@bors retry

new failures, probably intermittents

```
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/css/CSS2/floats-clear/clear-float-003.xht", "line": 5420, "action": "test_result", "expected": "PASS"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/selectors/pseudo-classes/checked.html", "line": 139160, "action": "test_result", "expected": "OK"}
```

**Comment by CYBAI** [comment]
@bors-servo retry

(I guess @Manishearth called wrong bors XD)

**Comment by Manishearth** [comment]
@bors-servo retry

 - no space on device

**Comment by Manishearth** [comment]
@bors-servo retry

 - font intermittent crash

**Pull request merged**